### PR TITLE
Add PerfDiagnosticListener

### PR DESCRIPTION
**Pull request by mikeharder**

### CODE REVIEW

File: PerfDiagnosticListener/PerfDiagnosticListener.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        private static long _requestsEnded = 0;
+        private static long _requestTicks = 0;
+
+        [DiagnosticName("Microsoft.AspNetCore.Hosting.BeginRequest")]
+        public void OnBeginRequest(HttpContext httpContext, long timestamp)
+        {
+            Interlocked.Increment(ref _requestsBegan);
+
+            var requestInfo = new RequestInfo { BeginTimestamp = timestamp };
+            httpContext.Features.Set(requestInfo);
```
Comment (davidfowl) [inline PerfDiagnosticListener/PerfDiagnosticListener.cs]: Don't use features here, use good old Request.Items

File: PerfDiagnosticListener/PerfDiagnosticListener.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        }
+
+        [DiagnosticName("Microsoft.AspNetCore.Hosting.EndRequest")]
+        public void OnEndRequest(HttpContext httpContext, long timestamp)
+        {
+            Interlocked.Increment(ref _requestsEnded);
```
Comment (davidfowl) [inline PerfDiagnosticListener/PerfDiagnosticListener.cs]: Interlocked.Decrement

File: PerfDiagnosticListener/PerfDiagnosticListener.cs (hunk 1 above)
Comment (mikeharder) [reply PerfDiagnosticListener/PerfDiagnosticListener.cs]: I will make this change, but then we should also change https://github.com/Microsoft/ApplicationInsights-aspnetcore/blob/develop/src/Microsoft.ApplicationInsights.AspNetCore/DiagnosticListeners/Implementation/HostingDiagnosticListener.cs in the same way.  Were we involved in the implementation of this?  @pakrym

Comment (mikeharder) [reply PerfDiagnosticListener/PerfDiagnosticListener.cs]: I assume you mean `HttpContext.Items`, since `HttpContext.Request.Items` does not exist.

Comment (pakrym) [reply PerfDiagnosticListener/PerfDiagnosticListener.cs]: @davidfowl why not?

Comment (mikeharder) [reply PerfDiagnosticListener/PerfDiagnosticListener.cs]: Reading the docs, I get the impression that `HttpContext.Features` is more intended for public stuff which may be used by multiple middlewares or app code, while `HttpContext.Items` is more intended for private stuff used by a single middleware or app code.  If true, this means `HttpContext.Items` is more appropriate in this case.  But I agree this is not clear.

**Pull request merged**